src/monogame/vulkan/MGG_Vulkan.cpp

```diff
@@ -0,0 +1,973 @@
+
+	VkViewport viewport;
+	VkRect2D scissor;
+	bool scissorDirty;
+
+	VkSurfaceKHR surface;
+	VkSwapchainKHR swapchain;
+	uint32_t swapchain_image_index;
+
+	uint64_t vertexBuffersDirty;
+	MGG_Buffer* vertexBuffers[8];
+
+	uint64_t currentTextureId;
+	uint64_t currentSamplerId;
+	MGG_Texture* textures[16];
+	MGG_SamplerState* samplers[16];
+	uint32_t textureSamplerDirty;
+
+	bool blendFactorDirty;
+	float blendFactor[4];
+
+	MGG_BlendState* blendState = nullptr;
+	MGG_DepthStencilState* depthStencilStateState = nullptr;
+	MGG_RasterizerState* rasterizerState = nullptr;
+	MGG_InputLayout* layout = nullptr;
+	MGG_Shader* shader[(mgint)MGShaderStage::Count] = { 0 };
+
+	VkPipelineCache pipelineCache;
+	//std::map<uint32_t, MGG_PipelineCache> pipelines;
+
+	std::map<uint32_t, MGG_BlendState*> blendStates;
+	std::map<uint32_t, MGG_RasterizerState*> rasterizerStates;
+	std::map<uint32_t, MGG_DepthStencilState*> depthStencilStates;
+
+	//MGG_TargetSet targets;
+	//std::map<uint32_t, MGG_TargetSetCache*> targetCache;
+
+	std::vector<MGG_Shader*> shaders;
+
+	//
+	bool pipelineStateDirty;
+	//MGG_PipelineState pipelineState;
+
+	MGG_Buffer* uniforms[2];
+	uint32_t uniformsDirty;
+
+	VkDescriptorSet descriptorSets[2];
+	uint32_t dynamicOffsets[2];
+
+	std::queue<MGG_Buffer*> destroyBuffers;
+	std::queue<MGG_Texture*> destroyTextures;
+	std::queue<MGG_BlendState*> destroyBlendStates;
+	std::queue<MGG_RasterizerState*> destroyRasterizerStates;
+	std::queue<MGG_DepthStencilState*> destroyDepthStencilStates;
+
+	MGG_Buffer* discarded;
+	MGG_Buffer* pending;
+	MGG_Buffer* free;
+
+	std::vector<MGG_Buffer*> all_buffers;
+	std::vector<MGG_Texture*> all_textures;
+};
+
+struct MGG_Buffer
+{
+	mgint length;
+};
+
+struct MGG_Texture
+{
+	MGTextureType type;
+	MGSurfaceFormat format;
+	mgint width;
+	mgint height;
+	mgint depth;
+	mgint mipmaps;
+	mgint slices;
+
+	mgbool isRenderTarget = false;
+	MGDepthFormat depthFormat = MGDepthFormat::None;
+	mgint multiSampleCount = 0;
+	MGRenderTargetUsage usage = MGRenderTargetUsage::PlatformContents;
+};
+
+struct MGG_InputLayout
+{
+	MGG_Shader* vertexShader;
+};
+
+struct MGG_Shader
+{
+	MGShaderStage stage;
+	mgbyte* bytecode;
+	mgint bytecodeSize;
+};
+
+struct MGG_BlendState
+{
+	MGG_BlendState_Info info[4];
+};
+
+struct MGG_DepthStencilState
+{
+	MGG_DepthStencilState_Info info;
+};
+
+struct MGG_RasterizerState
+{
+	MGG_RasterizerState_Info info;
+};
+
+struct MGG_SamplerState
+{
+	MGG_SamplerState_Info info;
+};
+
+
+struct MGG_System
+{
+	VkInstance instance;
+
+	std::vector<MGG_GraphicsAdapter*> adapters;
+};
+
+static MGG_System* mgg_system = nullptr;
+
+
+void MGG_EffectResource_GetBytecode(const char* name, mgbyte** bytecode, mgint* size)
+{
+	*bytecode = nullptr;
+	*size = 0;
+
+	// Get the handle of this DLL.
+	HMODULE module;
+	GetModuleHandleExA(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS, (LPCSTR)&MGG_EffectResource_GetBytecode, &module);
+
+	LPCSTR id = nullptr;
+
+	if (strcmp(name, "AlphaTestEffect") == 0)
```

Comment:
Should we consider making an enum for the stock effects instead of passing around string names here?

Reply:
Yes.